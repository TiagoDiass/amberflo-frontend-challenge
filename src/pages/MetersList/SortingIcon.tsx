import { Meter } from '@/types/meters';
import { SortMetersOptions } from '@/utils/sort-meters';
import { ArrowDown, ArrowUp, ArrowUpDown } from 'lucide-react';

type SortingIconProps = {
	field: keyof Meter;
	options: SortMetersOptions | null;
};

export function SortingIcon({ field, options }: SortingIconProps) {
	const classNames = 'size-4 text-zinc-800';

	if (!options || options.field !== field) {
		return <ArrowUpDown className={classNames} />;
	}

	return options.direction === 'ascending' ? (
		<ArrowDown className={classNames} />
	) : (
		<ArrowUp className={classNames} />
	);
}
