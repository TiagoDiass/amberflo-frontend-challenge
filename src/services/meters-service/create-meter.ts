import { Meter } from '@/types/meters';
import { HttpStatusCode, isAxiosError } from 'axios';
import { httpClient } from '../http-client';

export type CreateMeterParams = Pick<
	Meter,
	'api_name' | 'display_name' | 'active' | 'used_for_billing' | 'type'
>;

type CreateMeterResult = {
	successfullyCreated: boolean;
	errorMessage?: string;
};

export async function createMeter(
	params: CreateMeterParams,
): Promise<CreateMeterResult> {
	try {
		const response = await httpClient.post('/meters', {
			api_name: params.api_name,
			display_name: params.display_name,
			active: params.active,
			used_for_billing: params.used_for_billing,
			type: params.type,
		});

		return {
			successfullyCreated: response.status === HttpStatusCode.Created,
		};
	} catch (err: unknown) {
		console.error(err);

		return {
			successfullyCreated: false,
			errorMessage: isAxiosError(err)
				? err.response?.data.error
				: 'Unexpected error',
		};
	}
}
