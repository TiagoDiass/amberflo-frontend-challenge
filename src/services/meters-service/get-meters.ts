import { Meter } from '@/types/meters';
import { httpClient } from '../http-client';

export async function getMeters(): Promise<Meter[]> {
	const response = await httpClient.get('/meters');
	const meters = response.data;

	return meters;
}
