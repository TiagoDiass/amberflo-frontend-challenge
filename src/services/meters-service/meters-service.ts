import { Meter } from '@/types/meters';
import { HttpStatusCode, isAxiosError } from 'axios';
import { httpClient } from '../http-client';

async function getMeters(): Promise<Meter[]> {
	const response = await httpClient.get('/meters');
	const meters = response.data;

	return meters;
}

export type CreateMeterParams = Pick<
	Meter,
	'api_name' | 'display_name' | 'active' | 'used_for_billing' | 'type'
>;

type CreateMeterResult = {
	successfullyCreated: boolean;
	errorMessage?: string;
};

async function createMeter(
	params: CreateMeterParams,
): Promise<CreateMeterResult> {
	try {
		const response = await httpClient.post('/meters', {
			api_name: params.api_name,
			display_name: params.display_name,
			active: params.active,
			used_for_billing: params.used_for_billing,
			type: params.type,
		});

		return {
			successfullyCreated: response.status === HttpStatusCode.Created,
		};
	} catch (err: unknown) {
		console.error(err);

		return {
			successfullyCreated: false,
			errorMessage: isAxiosError(err)
				? err.response?.data.error
				: 'Unexpected error',
		};
	}
}

export const metersService = {
	getMeters,
	createMeter,
};
