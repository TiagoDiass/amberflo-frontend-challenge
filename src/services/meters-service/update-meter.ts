import { Meter } from '@/types/meters';
import { HttpStatusCode, isAxiosError } from 'axios';
import { httpClient } from '../http-client';

export type UpdateMeterParams = Pick<
	Meter,
	'api_name' | 'display_name' | 'active' | 'used_for_billing' | 'type'
>;

type UpdateMeterResult = {
	successfullyUpdated: boolean;
	errorMessage?: string;
};

export async function updateMeter(
	meterId: string,
	params: UpdateMeterParams,
): Promise<UpdateMeterResult> {
	try {
		const response = await httpClient.put(`/meters/${meterId}`, {
			api_name: params.api_name,
			display_name: params.display_name,
			active: params.active,
			used_for_billing: params.used_for_billing,
			type: params.type,
		});

		return {
			successfullyUpdated: response.status === HttpStatusCode.Created,
		};
	} catch (err: unknown) {
		console.error(err);

		return {
			successfullyUpdated: false,
			errorMessage: isAxiosError(err)
				? err.response?.data.error
				: 'Unexpected error',
		};
	}
}
