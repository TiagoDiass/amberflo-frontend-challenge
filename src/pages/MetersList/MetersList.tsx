import { Badge } from '@/components/Badge/Badge';
import { Button } from '@/components/ui/button';
import {
	Table,
	TableBody,
	TableCell,
	TableHead,
	TableHeader,
	TableRow,
} from '@/components/ui/table';
import { metersService } from '@/services/meters-service/meters-service';
import { useMetersStore } from '@/store/use-meters-store';
import { Meter } from '@/types/meters';
import { formatters } from '@/utils/formatters';
import { useEffect, useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';

type TableColumns = Array<{
	title: string;
	field: keyof Meter;
}>;

const tableColumns: TableColumns = [
	{ title: 'Display Name', field: 'display_name' },
	{ title: 'API Name', field: 'api_name' },
	{ title: 'Meter Type', field: 'type' },
	{ title: 'Status', field: 'active' },
	{ title: 'Used for billing', field: 'used_for_billing' },
	{ title: 'Modified', field: 'updated_time' },
];

export function MetersList() {
	const [meters, setMeters] = useState<Meter[]>([]);
	const setCurrentMeter = useMetersStore(
		(store) => store.actions.setCurrentMeter,
	);
	const navigate = useNavigate();

	function navigateToMeterDetails(meter: Meter) {
		setCurrentMeter(meter);
		navigate(`/meters/${meter.id}/details`);
	}

	useEffect(() => {
		async function fetchMeters() {
			const metersResponse = await metersService.getMeters();

			setMeters(metersResponse);
		}

		fetchMeters();
	}, []);

	return (
		<div className="max-w-screen-lg mx-auto px-6">
			<div className="flex justify-between items-center">
				<h1 className="font-semibold text-xl text-zinc-700">Your meters</h1>

				<Link to="/create-meter">
					<Button>Add new meter</Button>
				</Link>
			</div>

			<Table className="border border-zinc-200 mt-4">
				<TableHeader>
					<TableRow>
						{tableColumns.map((column) => (
							<TableHead key={column.field}>{column.title}</TableHead>
						))}
					</TableRow>
				</TableHeader>

				<TableBody>
					{meters.map((meter) => (
						<TableRow
							key={meter.id}
							onClick={() => navigateToMeterDetails(meter)}
							className="cursor-pointer"
						>
							<TableCell data-testid="meter-display-name">
								{meter.display_name}
							</TableCell>

							<TableCell className="italic">{meter.api_name}</TableCell>

							<TableCell className="italic">{meter.type}</TableCell>

							<TableCell>
								{meter.active ? (
									<Badge color="green">Active</Badge>
								) : (
									<Badge color="red">Inactive</Badge>
								)}
							</TableCell>

							<TableCell>{meter.used_for_billing ? 'Yes' : 'No'}</TableCell>

							<TableCell>{formatters.formatDate(meter.updated_time)}</TableCell>
						</TableRow>
					))}
				</TableBody>
			</Table>
		</div>
	);
}
