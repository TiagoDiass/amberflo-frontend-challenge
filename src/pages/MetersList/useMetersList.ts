import { metersService } from '@/services/meters-service/meters-service';
import { useMetersStore } from '@/store/use-meters-store';
import { Meter } from '@/types/meters';
import { SortMetersOptions, sortMeters } from '@/utils/sort-meters';
import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';

export function useMetersList() {
	const setCurrentMeter = useMetersStore(
		(store) => store.actions.setCurrentMeter,
	);
	const [meters, setMeters] = useState<Meter[]>([]);
	const [sortOptions, setSortOptions] = useState<SortMetersOptions | null>(
		null,
	);

	const sortedMeters = sortMeters(meters, sortOptions);
	const navigate = useNavigate();

	function navigateToMeterDetails(meter: Meter) {
		setCurrentMeter(meter);
		navigate(`/meters/${meter.id}/details`);
	}

	function handleSortMeters(field: keyof Meter) {
		if (sortOptions?.field === field) {
			const currentDirection = sortOptions.direction;
			const newDirection =
				currentDirection === 'ascending' ? 'descending' : 'ascending';

			setSortOptions({ field, direction: newDirection });
			return;
		}

		setSortOptions({
			field,
			direction: 'ascending',
		});
	}

	useEffect(() => {
		async function fetchMeters() {
			const metersResponse = await metersService.getMeters();

			setMeters(metersResponse);
		}

		fetchMeters();
	}, []);

	return {
		sortedMeters,
		sortOptions,
		handleSortMeters,
		navigateToMeterDetails,
	};
}
