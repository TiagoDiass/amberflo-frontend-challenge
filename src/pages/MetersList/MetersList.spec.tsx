import { metersService } from '@/services/meters-service/meters-service';
import { Meter } from '@/types/meters';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import { MetersList } from './MetersList';

function renderComponent() {
	const renderResult = render(
		<BrowserRouter>
			<MetersList />
		</BrowserRouter>,
	);

	return renderResult;
}

const metersMock: Meter[] = [
	{
		id: '3a309982-720c-49b6-a28a-4f5d9d415509',
		api_name: 'LOGINS',
		display_name: 'Logins',
		active: false,
		used_for_billing: false,
		updated_time: '2023-04-19T04:36:30.882Z',
		created_time: '2023-04-19T04:36:30.882Z',
		type: 'unique_count',
	},
	{
		id: '1634a14b-ecfa-405c-9113-5f71ae99b97a',
		api_name: 'API_CALLS',
		display_name: 'API Calls',
		active: true,
		used_for_billing: true,
		updated_time: '2023-06-07T20:09:22.933Z',
		created_time: '2023-04-19T00:57:55.628Z',
		type: 'sum',
	},
];

describe('Page: MetersList', () => {
	it('should render a list of meters correctly', async () => {
		vi.spyOn(metersService, 'getMeters').mockResolvedValue(metersMock);

		const { container } = renderComponent();

		await waitFor(() => {
			expect(screen.getAllByRole('row')).toHaveLength(3); // 3 rows -> 1 is the table header, 2 are from the table body
		});

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should sort the meters list correctly', async () => {
		vi.spyOn(metersService, 'getMeters').mockResolvedValue(metersMock);

		renderComponent();

		await waitFor(() => {
			expect(screen.getAllByRole('row')).toHaveLength(3);
		});

		const sortByDisplayNameButton = screen.getByRole('button', {
			name: 'Sort table by Display Name field',
		});
		await userEvent.click(sortByDisplayNameButton);

		const sortedCells = screen.getAllByTestId('meter-display-name');

		expect(sortedCells[0]).toHaveTextContent(metersMock[1].display_name);
		expect(sortedCells[1]).toHaveTextContent(metersMock[0].display_name);

		await userEvent.click(sortByDisplayNameButton);
		const reversedSorted = screen.getAllByTestId('meter-display-name');

		expect(reversedSorted[0]).toHaveTextContent(metersMock[0].display_name);
		expect(reversedSorted[1]).toHaveTextContent(metersMock[1].display_name);
	});
});
