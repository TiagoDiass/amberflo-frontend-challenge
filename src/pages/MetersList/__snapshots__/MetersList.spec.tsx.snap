// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Page: MetersList > should render a list of meters correctly 1`] = `
<div
  class="max-w-screen-lg mx-auto px-6"
>
  <div
    class="flex justify-between items-center"
  >
    <h1
      class="font-semibold text-xl text-zinc-700"
    >
      Your meters
    </h1>
    <a
      href="/create-meter"
    >
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-amber-600 text-zinc-50 hover:bg-amber-700 h-10 px-4 py-2"
      >
        Add new meter
      </button>
    </a>
  </div>
  <div
    class="relative w-full overflow-auto"
  >
    <table
      class="w-full caption-bottom text-sm border border-zinc-200 mt-4"
    >
      <thead
        class="[&_tr]:border-b"
      >
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
        >
          <th
            class="h-12 px-4 text-left align-middle font-semibold text-zinc-700 [&:has([role=checkbox])]:pr-0"
          >
            <div
              class="flex items-center gap-2"
            >
              Display Name
              <button
                class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-9 rounded-md px-3"
                title="Sort table by Display Name field"
              >
                <svg
                  class="lucide lucide-arrow-up-down size-4 text-zinc-800"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m21 16-4 4-4-4"
                  />
                  <path
                    d="M17 20V4"
                  />
                  <path
                    d="m3 8 4-4 4 4"
                  />
                  <path
                    d="M7 4v16"
                  />
                </svg>
              </button>
            </div>
          </th>
          <th
            class="h-12 px-4 text-left align-middle font-semibold text-zinc-700 [&:has([role=checkbox])]:pr-0"
          >
            <div
              class="flex items-center gap-2"
            >
              API Name
              <button
                class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-9 rounded-md px-3"
                title="Sort table by API Name field"
              >
                <svg
                  class="lucide lucide-arrow-up-down size-4 text-zinc-800"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m21 16-4 4-4-4"
                  />
                  <path
                    d="M17 20V4"
                  />
                  <path
                    d="m3 8 4-4 4 4"
                  />
                  <path
                    d="M7 4v16"
                  />
                </svg>
              </button>
            </div>
          </th>
          <th
            class="h-12 px-4 text-left align-middle font-semibold text-zinc-700 [&:has([role=checkbox])]:pr-0"
          >
            <div
              class="flex items-center gap-2"
            >
              Meter Type
              <button
                class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-9 rounded-md px-3"
                title="Sort table by Meter Type field"
              >
                <svg
                  class="lucide lucide-arrow-up-down size-4 text-zinc-800"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m21 16-4 4-4-4"
                  />
                  <path
                    d="M17 20V4"
                  />
                  <path
                    d="m3 8 4-4 4 4"
                  />
                  <path
                    d="M7 4v16"
                  />
                </svg>
              </button>
            </div>
          </th>
          <th
            class="h-12 px-4 text-left align-middle font-semibold text-zinc-700 [&:has([role=checkbox])]:pr-0"
          >
            <div
              class="flex items-center gap-2"
            >
              Status
              <button
                class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-9 rounded-md px-3"
                title="Sort table by Status field"
              >
                <svg
                  class="lucide lucide-arrow-up-down size-4 text-zinc-800"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m21 16-4 4-4-4"
                  />
                  <path
                    d="M17 20V4"
                  />
                  <path
                    d="m3 8 4-4 4 4"
                  />
                  <path
                    d="M7 4v16"
                  />
                </svg>
              </button>
            </div>
          </th>
          <th
            class="h-12 px-4 text-left align-middle font-semibold text-zinc-700 [&:has([role=checkbox])]:pr-0"
          >
            <div
              class="flex items-center gap-2"
            >
              Used for billing
              <button
                class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-9 rounded-md px-3"
                title="Sort table by Used for billing field"
              >
                <svg
                  class="lucide lucide-arrow-up-down size-4 text-zinc-800"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m21 16-4 4-4-4"
                  />
                  <path
                    d="M17 20V4"
                  />
                  <path
                    d="m3 8 4-4 4 4"
                  />
                  <path
                    d="M7 4v16"
                  />
                </svg>
              </button>
            </div>
          </th>
          <th
            class="h-12 px-4 text-left align-middle font-semibold text-zinc-700 [&:has([role=checkbox])]:pr-0"
          >
            <div
              class="flex items-center gap-2"
            >
              Modified
              <button
                class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-9 rounded-md px-3"
                title="Sort table by Modified field"
              >
                <svg
                  class="lucide lucide-arrow-up-down size-4 text-zinc-800"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m21 16-4 4-4-4"
                  />
                  <path
                    d="M17 20V4"
                  />
                  <path
                    d="m3 8 4-4 4 4"
                  />
                  <path
                    d="M7 4v16"
                  />
                </svg>
              </button>
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        class="[&_tr:last-child]:border-0"
      >
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted cursor-pointer"
        >
          <td
            class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
            data-testid="meter-display-name"
          >
            Logins
          </td>
          <td
            class="p-4 align-middle [&:has([role=checkbox])]:pr-0 italic"
          >
            LOGINS
          </td>
          <td
            class="p-4 align-middle [&:has([role=checkbox])]:pr-0 italic"
          >
            unique_count
          </td>
          <td
            class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
          >
            <div
              class="p-1 text-xs font-semibold text-zinc-50 text-center rounded-full opacity-95 bg-red-600"
            >
              Inactive
            </div>
          </td>
          <td
            class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
          >
            No
          </td>
          <td
            class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
          >
            4/19/2023, 1:36:30 AM
          </td>
        </tr>
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted cursor-pointer"
        >
          <td
            class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
            data-testid="meter-display-name"
          >
            API Calls
          </td>
          <td
            class="p-4 align-middle [&:has([role=checkbox])]:pr-0 italic"
          >
            API_CALLS
          </td>
          <td
            class="p-4 align-middle [&:has([role=checkbox])]:pr-0 italic"
          >
            sum
          </td>
          <td
            class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
          >
            <div
              class="p-1 text-xs font-semibold text-zinc-50 text-center rounded-full opacity-95 bg-green-600"
            >
              Active
            </div>
          </td>
          <td
            class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
          >
            Yes
          </td>
          <td
            class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
          >
            6/7/2023, 5:09:22 PM
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
